File: backend/schemas/mesh.py
# Copyright © 2026 AtomGradient
# 版权所有 © 2026 质子梯度（北京）科技有限公司

"""Pydantic schemas for /api/mesh/* endpoints (P0 transport security + event stream)."""

from __future__ import annotations

from typing import Any, Optional

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# QR payload  (mirrors Swift `QRPairingPayload`)
# ---------------------------------------------------------------------------


class QRPairingEndpoint(BaseModel):
    serviceType: str
    serviceName: str
    ipv4: Optional[str] = None
    port: int


class QRPairingPayloadModel(BaseModel):
    version: int = 1
    peerId: str
    displayName: str
    role: str = Field(..., description='"brain" | "sensor" | "peer"')
    endpoint: QRPairingEndpoint
    certFingerprint: str = Field(..., min_length=64, max_length=64)
    nonce: str
    expiresAt: int = Field(..., description="Unix seconds")


class QRPairingResponse(BaseModel):
    payload: QRPairingPayloadModel
    pin: str = Field(..., description="6-char base32 code displayed to user")
    ttl_seconds: int


class PinExchangeRequest(BaseModel):
    pin: str = Field(..., min_length=1, max_length=16)


class PairRequestBody(BaseModel):
    peer_id: str = Field(..., min_length=1, max_length=128)
    display_name: str = Field(..., min_length=1, max_length=128)
    fingerprint: str = Field(..., min_length=64, max_length=64)


class PairRequestResponse(BaseModel):
    pin: str
    nonce: str
    ttl_seconds: int


# ---------------------------------------------------------------------------
# Devices
# ---------------------------------------------------------------------------


class TrustedPeerModel(BaseModel):
    peer_id: str
    display_name: str
    fingerprint: str
    role: str
    paired_at: float
    last_seen_at: Optional[float] = None
    revoked: bool
    # Stage 3 P1.1: real-time sensor observations carried in keepalive ping (PingStats JSON dict).
    # nil = peer is a legacy client / has never reported. Fields include appId / eventStoreTotal /
    # factsClassified / factsRawUnclassified.
    last_stats: Optional[dict] = None
    last_stats_at: Optional[float] = None


class DevicesListResponse(BaseModel):
    local: dict = Field(..., description="This Mac's identity (peer_id, display_name, fingerprint, mesh_port, http_port, ipv4)")
    peers: list[TrustedPeerModel]
    pending: list[QRPairingPayloadModel]


class RevokeResponse(BaseModel):
    ok: bool = True
    peer_id: str


class DeleteResponse(BaseModel):
    ok: bool = True
    peer_id: str


class PeerStatusQuery(BaseModel):
    peer_id: Optional[str] = None
    fingerprint: Optional[str] = None


class PeerStatusResponse(BaseModel):
    known: bool
    trusted: bool
    revoked: bool
    peer_id: Optional[str] = None
    display_name: Optional[str] = None


class DeviceSnapshotIngestRequest(BaseModel):
    """HTTP mirror of the EdgeMesh `device_state_snapshot` op."""

    peer_id: Optional[str] = Field(
        default=None,
        description="Trusted peer id. Optional when snapshot.identity.peer_id is present.",
    )
    snapshot: dict[str, Any] = Field(..., description="DeviceLearningSnapshot JSON payload")
    source: str = Field(default="api", description="api | mesh | test")


class DeviceLifecycleStatus(BaseModel):
    schema_version: str
    phase: str
    phase_label: str
    ready_for_persona_chat: bool
    recommended_actions: list[str]
    reasons: list[str]


class DeviceSnapshotReceipt(BaseModel):
    schema_version: str
    peer_id: str
    source: str
    received_at: float
    snapshot_sha256: str
    lifecycle: DeviceLifecycleStatus


class DeviceSnapshotIngestResponse(BaseModel):
    ok: bool = True
    receipt: DeviceSnapshotReceipt


class DeviceSnapshotLatestResponse(BaseModel):
    ok: bool = True
    receipt: DeviceSnapshotReceipt
    snapshot: dict[str, Any]


class HaloCapsulePushRequest(BaseModel):
    """Request body for host-initiated Halo capsule package transfer."""

    peer_id: str = Field(..., min_length=1, max_length=128)
    neural_imprint_dir: Optional[str] = Field(
        default=None,
        min_length=1,
        description="Local directory containing neural_imprint.safetensors and sidecar files.",
    )
    artifact_id: Optional[str] = Field(
        default=None,
        min_length=1,
        description="Neural Imprint artifact registry id. Alternative to neural_imprint_dir.",
    )
    min_runtime_version: Optional[str] = Field(
        default=None,
        min_length=1,
        description="Minimum EdgeKit runtime version. Defaults to latest snapshot edge_kit_version.",
    )
    transfer_id: Optional[str] = Field(default=None, min_length=1, max_length=128)
    capsule_id: Optional[str] = Field(default=None, min_length=1, max_length=128)
    chunk_size: int = Field(
        default=1024 * 1024,
        ge=1,
        le=8 * 1024 * 1024,
        description="Binary chunk size for mesh transfer frames.",
    )


class HaloCapsulePushResponse(BaseModel):
    ok: bool = True
    peer_id: str
    transfer_id: str
    capsule_id: str
    base_model_id: str
    min_runtime_version: str
    artifact_sha256: str
    artifact_file_count: int
    frame_count: int
    payload_bytes: int
    chunk_size: int


class HaloCapsuleApplyStatusIngestRequest(BaseModel):
    """HTTP mirror of the EdgeMesh `halo_capsule_apply_status` op."""

    peer_id: str = Field(..., min_length=1, max_length=128)
    payload: dict[str, Any] = Field(..., description="HaloCapsuleApplyStatusPayload JSON")
    source: str = Field(default="api", description="api | mesh | test")


class HaloCapsuleApplyStatusReceipt(BaseModel):
    schema_version: str
    peer_id: str
    source: str
    received_at: float
    apply_status_sha256: str
    transfer_id: str
    capsule_id: str
    status: str
    artifact_sha256: Optional[str] = None
    canonical_sha256: Optional[str] = None
    runtime_version: Optional[str] = None
    prefix_token_count: Optional[int] = None
    applied_at_unix_seconds: Optional[float] = None
    error_code: Optional[str] = None


class HaloCapsuleTransferAckReceipt(BaseModel):
    schema_version: str
    peer_id: str
    source: str
    received_at: float
    ack_sha256: str
    ack_kind: str
    transfer_id: str
    accepted: bool
    reason: Optional[str] = None
    canonical_sha256: Optional[str] = None


class HaloCapsuleApplyStatusIngestResponse(BaseModel):
    ok: bool = True
    receipt: HaloCapsuleApplyStatusReceipt


class HaloCapsuleApplyStatusLatestResponse(BaseModel):
    ok: bool = True
    receipt: HaloCapsuleApplyStatusReceipt
    payload: dict[str, Any]


class HaloCapsuleCoordinatorPlanResponse(BaseModel):
    ok: bool = True
    schema_version: str
    peer_id: str
    connected: bool
    action: dict[str, Any]
    snapshot_sha256: Optional[str] = None
    lifecycle: Optional[DeviceLifecycleStatus] = None
    selected_model_id: Optional[str] = None
    load_state: Optional[str] = None
    data_readiness: Optional[str] = None
    learning: Optional[dict[str, Any]] = None
    artifact_count: Optional[int] = None
    matched_artifact: Optional[dict[str, Any]] = None
    last_apply_status: Optional[HaloCapsuleApplyStatusReceipt] = None
    last_transfer_ack: Optional[HaloCapsuleTransferAckReceipt] = None


class HaloCapsuleAutomationPreviewEntry(BaseModel):
    peer_id: str
    display_name: str
    connected: bool
    would_push: bool
    action: dict[str, Any]
    plan: dict[str, Any]


class HaloCapsuleAutomationPreviewResponse(BaseModel):
    ok: bool = True
    schema_version: str
    dry_run: bool
    peer_count: int
    candidate_count: int
    skipped_revoked_count: int
    entries: list[HaloCapsuleAutomationPreviewEntry]


class HaloCapsuleAutomationRunRequest(BaseModel):
    dry_run: bool = Field(
        default=True,
        description="true previews only; false pushes selected candidates once.",
    )
    peer_ids: list[str] = Field(
        default_factory=list,
        description="Required when dry_run=false. Limits execution to explicit peers.",
    )
    max_pushes: int = Field(
        default=1,
        ge=1,
        le=20,
        description="Upper bound for one run. Defaults to one device.",
    )
    chunk_size: int = Field(
        default=1024 * 1024,
        ge=1,
        le=8 * 1024 * 1024,
        description="Binary chunk size for mesh transfer frames.",
    )


class HaloCapsuleAutomationRunResult(BaseModel):
    peer_id: str
    display_name: Optional[str] = None
    dry_run: bool
    would_push: bool
    action_kind: str
    status: str
    push: Optional[HaloCapsulePushResponse] = None
    error: Optional[str] = None


class HaloCapsuleAutomationRunReceipt(BaseModel):
    schema_version: str
    run_id: str
    source: str
    received_at: float
    run_sha256: str
    dry_run: bool
    attempted_count: int
    pushed_count: int
    peer_ids: list[str]


class HaloCapsuleAutomationRunResponse(BaseModel):
    ok: bool = True
    dry_run: bool
    attempted_count: int
    pushed_count: int
    preview: dict[str, Any]
    results: list[HaloCapsuleAutomationRunResult]
    receipt: Optional[HaloCapsuleAutomationRunReceipt] = None


class HaloCapsuleAutomationRunLatestResponse(BaseModel):
    ok: bool = True
    receipt: HaloCapsuleAutomationRunReceipt
    request: dict[str, Any]
    response: dict[str, Any]


# ---------------------------------------------------------------------------
# Status
# ---------------------------------------------------------------------------


class MeshStatusResponse(BaseModel):
    transport_running: bool
    discovery_running: bool
    peer_id: str
    fingerprint: str
    mesh_port: int
    http_port: int
    ipv4: Optional[str] = None
    peers_count: int
    pending_count: int


# ---------------------------------------------------------------------------
# Events
# ---------------------------------------------------------------------------


class EventRecord(BaseModel):
    id: str
    timestamp: float                      # Unix seconds (more convenient for frontend / training pipeline)
    app_id: str
    event_type: str
    tags: list[str]
    source_peer_id: Optional[str] = None
    payload_size: int
    payload_b64: Optional[str] = None


class EventsListResponse(BaseModel):
    total: int                            # Total matches for current filter
    returned: int                         # Count returned in this response (subject to limit/offset)
    events: list[EventRecord]


class EventStatsResponse(BaseModel):
    total_events: int
    total_bytes: int
    oldest_timestamp: Optional[float] = None
    newest_timestamp: Optional[float] = None
    per_type: dict[str, int]
    per_source_peer: dict[str, int]

File: frontend/src/api/mesh.ts
// Copyright © 2026 AtomGradient
// 版权所有 © 2026 质子梯度（北京）科技有限公司

/**
 * EdgeMesh REST API client (P0 transport security + event stream).
 *
 * Mirrors backend/schemas/mesh.py. Types use camelCase fields from Swift
 * `QRPairingPayload` where relevant, snake_case for local-only structures.
 */

import client from './client';

// ---------------------------------------------------------------------------
// QR pairing payload (wire-compat with Swift QRPairingPayload)
// ---------------------------------------------------------------------------

export interface QRPairingEndpoint {
  serviceType: string;
  serviceName: string;
  ipv4: string | null;
  port: number;
}

export interface QRPairingPayload {
  version: number;
  peerId: string;
  displayName: string;
  role: 'brain' | 'sensor' | 'peer';
  endpoint: QRPairingEndpoint;
  certFingerprint: string;
  nonce: string;
  expiresAt: number;          // Unix seconds
}

export interface QRPairingResponse {
  payload: QRPairingPayload;
  pin: string;                // 6-char base32
  ttl_seconds: number;
}

// ---------------------------------------------------------------------------
// Devices
// ---------------------------------------------------------------------------

export interface LocalIdentity {
  peer_id: string;
  display_name: string;
  fingerprint: string;
  mesh_port: number;
  http_port: number;
  ipv4: string | null;
}

export interface TrustedPeer {
  peer_id: string;
  display_name: string;
  fingerprint: string;
  role: string;
  paired_at: number;            // Unix seconds
  last_seen_at: number | null;
  revoked: boolean;
}

export interface DevicesListResponse {
  local: LocalIdentity;
  peers: TrustedPeer[];
  pending: QRPairingPayload[];
}

export interface MeshStatus {
  transport_running: boolean;
  discovery_running: boolean;
  peer_id: string;
  fingerprint: string;
  mesh_port: number;
  http_port: number;
  ipv4: string | null;
  peers_count: number;
  pending_count: number;
}

// ---------------------------------------------------------------------------
// Events (training pipeline view)
// ---------------------------------------------------------------------------

export interface EventRecord {
  id: string;
  timestamp: number;             // Unix seconds
  app_id: string;
  event_type: string;
  tags: string[];
  source_peer_id: string | null;
  payload_size: number;
  payload_b64?: string;
}

export interface EventsListResponse {
  total: number;
  returned: number;
  events: EventRecord[];
}

export interface EventStats {
  total_events: number;
  total_bytes: number;
  oldest_timestamp: number | null;
  newest_timestamp: number | null;
  per_type: Record<string, number>;
  per_source_peer: Record<string, number>;
}

// ---------------------------------------------------------------------------
// Halo capsule / Neural Imprint artifact distribution
// ---------------------------------------------------------------------------

export interface NeuralImprintArtifactSource {
  valid: boolean;
  artifact_id?: string;
  capsule_id?: string;
  transfer_id_hint?: string;
  artifact_dir: string;
  artifact_path: string;
  sidecar_path: string;
  base_model_id?: string;
  model_architecture?: string;
  model_family?: string;
  hidden_size?: number;
  layer_count?: number;
  prefix_token_count?: number;
  tool_schema_sha256?: string;
  profile_body_sha256?: string;
  cache_backend?: string;
  cache_backend_version?: string;
  cache_topology_sha256?: string;
  neural_imprint_sha256?: string | null;
  artifact_sha256?: string;
  total_bytes?: number;
  file_count?: number;
  created_at?: number | string | null;
  mtime?: number | null;
  error?: string;
}

export interface NeuralImprintArtifactRegistryResponse {
  ok: boolean;
  schema_version: string;
  roots: string[];
  artifact_count: number;
  artifacts: NeuralImprintArtifactSource[];
}

export interface PersonaSourceLatestResponse {
  ok: boolean;
  receipt: {
    peer_id: string;
    source_id: string;
    source_kind: string;
    base_model_id: string;
    tool_schema_sha256: string;
    profile_body_sha256?: string | null;
    rpp_run_id?: string | null;
    received_at_ms?: number | null;
  };
  payload: Record<string, unknown>;
}

export interface NeuralImprintGenerateRequest {
  peer_id: string;
  model_dir: string;
  model_id?: string;
  validate_restore?: boolean;
}

export interface NeuralImprintGenerationJob {
  schema_version: string;
  job_id: string;
  status: 'queued' | 'running' | 'succeeded' | 'failed' | string;
  peer_id: string;
  source_id: string;
  model_dir: string;
  model_id: string;
  output_dir: string;
  validate_restore: boolean;
  created_at: number;
  updated_at: number;
  started_at?: number | null;
  completed_at?: number | null;
  result?: {
    schema_version: string;
    artifact_dir: string;
    artifact_path: string;
    metadata_path: string;
    receipt_path: string;
    peer_id: string;
    source_id: string;
    model_id: string;
    prefix_token_count: number;
    artifact_sha256: string;
    tool_schema_sha256: string;
    profile_body_sha256: string;
  } | null;
  error?: {
    code: string;
    message: string;
    retryable?: boolean;
    details?: Record<string, unknown>;
  } | null;
}

export interface NeuralImprintGenerateJobResponse {
  ok: boolean;
  job: NeuralImprintGenerationJob;
}


export interface HaloCapsulePushRequest {
  peer_id: string;
  neural_imprint_dir?: string;
  artifact_id?: string;
  min_runtime_version?: string;
  transfer_id?: string;
  capsule_id?: string;
  chunk_size?: number;
}

export interface HaloCapsulePushResponse {
  ok: boolean;
  peer_id: string;
  transfer_id: string;
  capsule_id: string;
  base_model_id: string;
  min_runtime_version: string;
  artifact_sha256: string;
  artifact_file_count: number;
  frame_count: number;
  payload_bytes: number;
  chunk_size: number;
}

export interface HaloCapsuleApplyStatusReceipt {
  schema_version: string;
  peer_id: string;
  source: string;
  received_at: number;
  apply_status_sha256: string;
  transfer_id: string;
  capsule_id: string;
  status: 'received' | 'applied' | 'failed' | string;
  artifact_sha256?: string | null;
  canonical_sha256?: string | null;
  runtime_version?: string | null;
  prefix_token_count?: number | null;
  applied_at_unix_seconds?: number | null;
  error_code?: string | null;
}

export interface HaloCapsuleTransferAckReceipt {
  schema_version: string;
  peer_id: string;
  source: string;
  received_at: number;
  ack_sha256: string;
  ack_kind: string;
  transfer_id: string;
  accepted: boolean;
  reason?: string | null;
  canonical_sha256?: string | null;
}

export interface HaloCapsuleApplyStatusLatestResponse {
  ok: boolean;
  receipt: HaloCapsuleApplyStatusReceipt;
  payload: Record<string, unknown>;
}

export interface HaloCapsuleCoordinatorAction {
  kind: string;
  can_push: boolean;
  requires_user_confirmation: boolean;
  label: string;
  reasons: string[];
  push_request?: HaloCapsulePushRequest;
}

export interface HaloCapsuleCoordinatorPlanResponse {
  ok: boolean;
  schema_version: string;
  peer_id: string;
  connected: boolean;
  action: HaloCapsuleCoordinatorAction;
  snapshot_sha256?: string | null;
  lifecycle?: Record<string, unknown> | null;
  selected_model_id?: string | null;
  load_state?: string | null;
  data_readiness?: string | null;
  learning?: Record<string, unknown> | null;
  artifact_count?: number | null;
  matched_artifact?: NeuralImprintArtifactSource | null;
  last_apply_status?: HaloCapsuleApplyStatusReceipt | null;
  last_transfer_ack?: HaloCapsuleTransferAckReceipt | null;
}

export interface HaloCapsuleAutomationPreviewEntry {
  peer_id: string;
  display_name: string;
  connected: boolean;
  would_push: boolean;
  action: HaloCapsuleCoordinatorAction;
  plan: HaloCapsuleCoordinatorPlanResponse;
}

export interface HaloCapsuleAutomationPreviewResponse {
  ok: boolean;
  schema_version: string;
  dry_run: boolean;
  peer_count: number;
  candidate_count: number;
  skipped_revoked_count: number;
  entries: HaloCapsuleAutomationPreviewEntry[];
}

export interface HaloCapsuleAutomationRunRequest {
  dry_run?: boolean;
  peer_ids?: string[];
  max_pushes?: number;
  chunk_size?: number;
}

export interface HaloCapsuleAutomationRunResult {
  peer_id: string;
  display_name?: string | null;
  dry_run: boolean;
  would_push: boolean;
  action_kind: string;
  status: 'candidate' | 'skipped' | 'pushed' | 'failed' | string;
  push?: HaloCapsulePushResponse | null;
  error?: string | null;
}

export interface HaloCapsuleAutomationRunResponse {
  ok: boolean;
  dry_run: boolean;
  attempted_count: number;
  pushed_count: number;
  preview: HaloCapsuleAutomationPreviewResponse;
  results: HaloCapsuleAutomationRunResult[];
}

export interface JointInferenceHistoryItem {
  schema_version: string;
  request_id: string;
  conversation_id?: string | null;
  latest_request_id?: string | null;
  parent_request_id?: string | null;
  source?: string | null;
  peer_id: string;
  status: string;
  accepted_at: number;
  completed_at?: number | null;
  duration_seconds?: number | null;
  model_id?: string | null;
  max_tokens?: number | null;
  temperature?: number | null;
  enable_thinking?: boolean | null;
  use_neural_imprint?: boolean | null;
  neural_imprint_artifact_id?: string | null;
  neural_imprint_prefix_token_count?: number | null;
  route_reason?: string | null;
  messages_count?: number | null;
  messages?: Array<{ role: string; content: string }>;
  prompt_preview?: string | null;
  output_preview?: string | null;
  full_text?: string | null;
  error?: string | null;
  total_tokens?: number | null;
  tokens_per_sec?: number | null;
  last_event_type?: string | null;
  last_sequence?: number | null;
  token_events?: number | null;
}

export interface JointInferenceHistoryResponse {
  ok: boolean;
  schema_version: string;
  count: number;
  items: JointInferenceHistoryItem[];
}

export interface JointInferenceHistoryDetailResponse {
  ok: boolean;
  item: JointInferenceHistoryItem;
}

// ---------------------------------------------------------------------------
// Endpoints
// ---------------------------------------------------------------------------

export const getMeshStatus = () =>
  client.get<MeshStatus>('/mesh/status').then((r) => r.data);

export const listDevices = () =>
  client.get<DevicesListResponse>('/mesh/devices').then((r) => r.data);

export const createPairing = () =>
  client.post<QRPairingResponse>('/mesh/pair/qr').then((r) => r.data);

export const revokePeer = (peerId: string) =>
  client.post<{ ok: boolean; peer_id: string }>(
    `/mesh/devices/${encodeURIComponent(peerId)}/revoke`
  ).then((r) => r.data);

export const deletePeer = (peerId: string) =>
  client.delete<{ ok: boolean; peer_id: string }>(
    `/mesh/devices/${encodeURIComponent(peerId)}`
  ).then((r) => r.data);

export const approvePairRequest = (nonce: string) =>
  client.post<{ ok: boolean; nonce: string }>(
    `/mesh/pair/approve/${encodeURIComponent(nonce)}`
  ).then((r) => r.data);

export const listEvents = (params?: {
  tags?: string[];
  app_id?: string;
  event_type?: string;
  source_peer_id?: string;
  since?: number;
  until?: number;
  limit?: number;
  offset?: number;
  include_payload?: boolean;
}) =>
  client
    .get<EventsListResponse>('/mesh/events', { params })
    .then((r) => r.data);

export const getEventStats = () =>
  client.get<EventStats>('/mesh/events/stats').then((r) => r.data);

export const getJointInferenceHistory = (params?: { peer_id?: string; limit?: number }) =>
  client
    .get<JointInferenceHistoryResponse>('/mesh/joint_inference/history', { params })
    .then((r) => r.data);

export const getJointInferenceHistoryItem = (requestId: string) =>
  client
    .get<JointInferenceHistoryDetailResponse>(`/mesh/joint_inference/history/${requestId}`)
    .then((r) => r.data);

export const deleteJointInferenceHistoryItem = (requestId: string) =>
  client
    .delete<{ ok: boolean; request_id: string }>(`/mesh/joint_inference/history/${encodeURIComponent(requestId)}`)
    .then((r) => r.data);

export type JointInferenceStreamEvent = {
  schema_version: string;
  request_id: string;
  type: 'accepted' | 'status' | 'token' | 'complete' | 'error' | 'cancelled';
  sequence?: number;
  message?: string;
  token?: string;
  full_text?: string;
  total_tokens?: number;
  tokens_per_sec?: number;
  error?: string;
  use_neural_imprint?: boolean;
  neural_imprint_artifact_id?: string;
  neural_imprint_prefix_token_count?: number;
};

export async function streamJointInferenceContinue(
  requestId: string,
  payload: {
    messages?: Array<{ role: string; content: string }>;
    message?: string;
    max_tokens?: number;
    temperature?: number;
    enable_thinking?: boolean;
    use_neural_imprint?: boolean;
  },
  onEvent: (event: JointInferenceStreamEvent) => void,
) {
  const response = await fetch(`/api/mesh/joint_inference/history/${requestId}/continue`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });
  if (!response.ok) {
    throw new Error(await response.text());
  }
  if (!response.body) {
    throw new Error('Streaming response body is unavailable');
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = '';
  while (true) {
    const { value, done } = await reader.read();
    if (done) break;
    buffer += decoder.decode(value, { stream: true });
    const lines = buffer.split('\n');
    buffer = lines.pop() ?? '';
    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed) continue;
      onEvent(JSON.parse(trimmed) as JointInferenceStreamEvent);
    }
  }
  buffer += decoder.decode();
  const trimmed = buffer.trim();
  if (trimmed) {
    onEvent(JSON.parse(trimmed) as JointInferenceStreamEvent);
  }
}

export const listNeuralImprintArtifacts = (includeInvalid = false) =>
  client
    .get<NeuralImprintArtifactRegistryResponse>('/neural_imprint/artifacts', {
      params: { include_invalid: includeInvalid },
    })
    .then((r) => r.data);

export const getLatestPersonaSource = (peerId: string) =>
  client
    .get<PersonaSourceLatestResponse>(
      `/neural_imprint/sources/${encodeURIComponent(peerId)}/latest`,
    )
    .then((r) => r.data);

export const generateNeuralImprint = (request: NeuralImprintGenerateRequest) =>
  client.post<NeuralImprintGenerateJobResponse>('/neural_imprint/generate', request).then((r) => r.data);

export const getNeuralImprintGenerationJob = (jobId: string) =>
  client
    .get<NeuralImprintGenerateJobResponse>(`/neural_imprint/generate/${encodeURIComponent(jobId)}`)
    .then((r) => r.data);


export const pushHaloCapsule = (request: HaloCapsulePushRequest) =>
  client.post<HaloCapsulePushResponse>('/mesh/halo_capsules/push', request).then((r) => r.data);

export const getHaloCapsuleApplyStatus = (peerId: string) =>
  client
    .get<HaloCapsuleApplyStatusLatestResponse>(
      `/mesh/devices/${encodeURIComponent(peerId)}/halo_capsules/apply_status/latest`,
    )
    .then((r) => r.data);

export const getHaloCapsulePlan = (peerId: string) =>
  client
    .get<HaloCapsuleCoordinatorPlanResponse>(
      `/mesh/devices/${encodeURIComponent(peerId)}/halo_capsules/plan`,
    )
    .then((r) => r.data);

export const getHaloCapsuleAutomationPreview = () =>
  client
    .get<HaloCapsuleAutomationPreviewResponse>('/mesh/halo_capsules/automation/preview')
    .then((r) => r.data);

export const runHaloCapsuleAutomation = (request: HaloCapsuleAutomationRunRequest) =>
  client
    .post<HaloCapsuleAutomationRunResponse>('/mesh/halo_capsules/automation/run', request)
    .then((r) => r.data);
